spark/v3.4/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/optimizer/FoldConstantSystemFunction.scala

```diff
@@ -0,0 +1,53 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.spark.sql.catalyst.optimizer
+
+import org.apache.iceberg.spark.functions.UnboundFunctionWrapper.SystemFunctionWrapper
+import org.apache.spark.sql.catalyst.expressions.Alias
+import org.apache.spark.sql.catalyst.expressions.ApplyFunctionExpression
+import org.apache.spark.sql.catalyst.expressions.EmptyRow
+import org.apache.spark.sql.catalyst.expressions.Literal
+import org.apache.spark.sql.catalyst.plans.logical.LocalRelation
+import org.apache.spark.sql.catalyst.plans.logical.LogicalPlan
+import org.apache.spark.sql.catalyst.plans.logical.Project
+import org.apache.spark.sql.catalyst.rules.Rule
+
+/**
+ * Folding those system function that applies a constant value.
+ */
+object FoldConstantSystemFunction extends Rule[LogicalPlan] {
+
+  override def apply(plan: LogicalPlan): LogicalPlan = plan transform {
+    case logicalPlan: LogicalPlan =>
+      logicalPlan transformExpressions {
+        case applyFunc: ApplyFunctionExpression
+             if applyFunc.function.isInstanceOf[SystemFunctionWrapper[_]] =>
+          val fakeProjectList = applyFunc.children.zipWithIndex.map(pair => Alias(pair._1, s"fake_${pair._2}")())
+          val fakePlan = Project(fakeProjectList, LocalRelation(Nil))
+          val foldedExprs =
+            ConstantFolding.apply(fakePlan).asInstanceOf[Project].projectList.map(_.asInstanceOf[Alias].child)
```

Comment:
Why is this fake plan needed? Wouldn't Spark already call `ConstantFolding` on both `ApplyFunctionExpression` and `StaticInvoke`?

Reply:
Looks like this is because `ApplyFunctionExpression` doesn't override `foldable`.

I think we could fix that by adding this upstream:
```scala
  override def foldable: Boolean = deterministic && children.forall(_.foldable)
```

`StaticInvoke` already has `foldable` implemented by its `InvokeLike` parent:

```scala
  override def foldable: Boolean =
    children.forall(_.foldable) && deterministic && trustedSerializable(dataType)
```

That's basically the same, except that `trustedSerializable` checks that there isn't an `ObjectType` in `dataType`, which is the output type. I'm fairly sure that will be `true`, so I'm not sure why `StaticInvoke` is not already folded. We'll have to do some debugging to track it down.

Reply:
> Why is this fake plan needed?

`ConstantFolding` takes a logic plan as input.

> Wouldn't Spark already call ConstantFolding on both ApplyFunctionExpression and StaticInvoke?

Yeah, `ApplyFunctionExpression` can not be folded because it is not `foldable`.

> so I'm not sure why StaticInvoke is not already folded.

`StaticInvoke` can be folded. Here the rule is just for `ApplyFunctionExpression` to address it unfoldable problem.

> I think we could fix that by adding this upstream:

Yeah, I think we should patch to spark. I will submit a PR for it. Here is a temporary fix.